storage: make RaftTombstoneKey unreplicated

tbg opened the pull request:
RaftTombstoneKey was accidentally made a replicated key when it was first
introduced, a problem we first realized existed when it was [included in
snapshots].

At the time, we included workarounds to skip this key in various places
(snapshot application, consistency checker) but of course we have failed to
insert further hacks of the same kind elsewhere since (the one that prompting
this PR being the stats recomputation on splits, which I'm looking into as part
of #20181 -- unfortunately this commit doesn't seem to pertain to that problem)

It feels sloppy that we didn't follow through back then, but luckily the damage
appears to be limited; it is likely that the replicated existence of this key
results in MVCCStats SysBytes inconsistencies, but as it happens, these stats
are [already] [very] [inconsistent].

This commit does a few things:

- renames the old tombstone key to `RaftIncorrectLegacyTombstoneKey`
- introduces a (correctly unreplicated) `RaftTombstoneKey`
- introduces a migration that rewrites all legacy keys early in the node
  start sequence; as a result, a node running this commit or later will
  never see any legacy keys, except when they come in through a snapshot
- when applying a snapshot, forcibly delete any legacy tombstone contained
  within.

`RaftIncorrectLegacyTombstoneKey` can be purged from the codebase in binaries
post v2.0, as at that point all peers have version at least v2.0 (which has this
commit and will never send legacy tombstones in snapshots).

Since sending legacy keys in snapshots was never intended (and isn't relied
upon), none of this needs a cluster version migration.

Fixes #12154.

Release note: None

[included in snapshots]: https://github.com/cockroachdb/cockroach/pull/12131
[already]: https://github.com/cockroachdb/cockroach/issues/20554
[very]: https://github.com/cockroachdb/cockroach/pull/20996
[inconsistent]: https://github.com/cockroachdb/cockroach/pull/21070

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/21120)
<!-- Reviewable:end -->

tschottdorf commented on pkg/storage/store.go:
```diff
@@ -1047,6 +1047,35 @@ func IterateRangeDescriptors(
 	return err
 }
 
+func migrateLegacyTombstone(ctx context.Context, eng engine.Engine, rangeID roachpb.RangeID) error {
```
Ah, I think I need to do more here -- most tombstones aren't tied to a RangeID.

petermattis commented:
While I filed #12154, I was never intimately familiar with Raft tombstone keys. @bdarnell or someone else should give this PR a thorough look.

---

Review status: 0 of 6 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica_command.go, line 917 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21120#-L1YGAXIUDnGenOczU0f:-L1YGAXIUDnGenOczU0g:bmwp1nr) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3bee2ae33e90c6cad3770d73268fe2d8e4e4a3e/pkg/storage/replica_command.go#L917)):*
> ```Go
> 			break
> 		}
> 		// TODO(tschottdorf): looks like we'd want to use Unsafe{Key,Value} here when `snapshot == nil` (the common case).
> ```

Yes, we should do that. You could `s/key/unsafeKey/g` and `s/value/unsafeValue/g` and then use `iter.Key().Key` and `iter.Value()` inside of the `snapshot != nil` condition.

---

*[pkg/storage/store.go, line 1139 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21120#-L1YISxHLMRHtY91x6DI:-L1YISxHLMRHtY91x6DJ:betd2zf) ([raw file](https://github.com/cockroachdb/cockroach/blob/3277ae571b5c0ccc031e253b2744145393f8d1c9/pkg/storage/store.go#L1139)):*
> ```Go
> 			return false, err
> 		}
> 		err := batch.Commit(true /* sync */)
> ```

Is `sync == true` necessary? This will be a lot faster without it.

---

*[pkg/storage/store.go, line 1203 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21120#-L1YIH5gv2DzgEDA46V7:-L1YIH5gv2DzgEDA46V8:b-5ffyka) ([raw file](https://github.com/cockroachdb/cockroach/blob/3277ae571b5c0ccc031e253b2744145393f8d1c9/pkg/storage/store.go#L1203)):*
> ```Go
> 	if err := migrateLegacyTombstones(ctx, s.engine); err != nil {
> 		return errors.Wrap(err, "migrating legacy tombstones")
> 	}
> ```

Didn't we use to have some other hook to call migrations? I can't find it looking just now. Perhaps it was removed at some point.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21120#-:-L1YIie3ohbac98nwXpE:bmbonmq)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
TFTR and PTAL!
Agreed that @bdarnell should give a thorough review as well.

---

Review status: 0 of 7 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/storage/replica_command.go, line 917 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21120#-L1YGAXIUDnGenOczU0f:-L1YcnEqKSOaGWnxdZVe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3bee2ae33e90c6cad3770d73268fe2d8e4e4a3e/pkg/storage/replica_command.go#L917)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Yes, we should do that. You could `s/key/unsafeKey/g` and `s/value/unsafeValue/g` and then use `iter.Key().Key` and `iter.Value()` inside of the `snapshot != nil` condition.
</blockquote></details>

Done.

---

*[pkg/storage/store.go, line 1139 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21120#-L1YISxHLMRHtY91x6DI:-L1YZdMkITHkHY36M0S4:b-6tlgkz) ([raw file](https://github.com/cockroachdb/cockroach/blob/3277ae571b5c0ccc031e253b2744145393f8d1c9/pkg/storage/store.go#L1139)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Is `sync == true` necessary? This will be a lot faster without it.
</blockquote></details>

I think it's reasonable to persist these changes once (or I won't be able to ever guarantee that these keys are gone), but I refactored this to sync only once at the end.

---

*[pkg/storage/store.go, line 1203 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/21120#-L1YIH5gv2DzgEDA46V7:-L1YbA4pY4AILv4Lfo3c:b-r4vfje) ([raw file](https://github.com/cockroachdb/cockroach/blob/3277ae571b5c0ccc031e253b2744145393f8d1c9/pkg/storage/store.go#L1203)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Didn't we use to have some other hook to call migrations? I can't find it looking just now. Perhaps it was removed at some point.
</blockquote></details>

Yeah, I remembered that too, but I *think* they've since been removed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21120#-:-L1Ycr50pXi1ootutSlQ:b-gj8rqs)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 7 files at r1, 2 of 4 files at r2, 1 of 1 files at r3, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/storage/store.go, line 1053 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/21120#-L1ZP5GFcXD1BZQXA9vd:-L1ZObmX99IqAKI2usGi:bjv3nk0) ([raw file](https://github.com/cockroachdb/cockroach/blob/79968b95c10bcd189226e7fcb219dd22d3c08da7/pkg/storage/store.go#L1053)):*
> ```Go
> 
> 		if !unsafeKey.Key.Equal(mvccKey.Key) {
> 			rangeID++
> ```

This looks like it might skip keys: If we have

- /Local/Range/1/u/Tombstone
- /Local/Range/3/u/Something
- /Local/Range/3/u/Tombstone

then when we try to seek to 2/u/Tombstone, we'll go to 3/u/Something instead. We have to seek to 3/u/Tombstone, but this increment will send us to 4/u/Tombstone instead. We should only change rangeID once, either to increment it or to set it to the one that we got by decoding.

(This may actually be safe since we should only have a tombstone when nothing else exists for the range, but I'd feel better with this change)

---

*[pkg/storage/store.go, line 1057 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/21120#-L1ZQ5TBMtNkEmqRYXCh:-L1ZQ5TBMtNkEmqRYXCi:b86s34y) ([raw file](https://github.com/cockroachdb/cockroach/blob/79968b95c10bcd189226e7fcb219dd22d3c08da7/pkg/storage/store.go#L1057)):*
> ```Go
> 		}
> 
> 		ok, err := engine.MVCCGetProto(
> ```

We should explicitly clear `msg` on each cycle; otherwise bits from one iteration could bleed into the next (not an issue for the trivial RaftTruncatedState but could matter for other protos)

---

*[pkg/storage/store.go, line 1067 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/21120#-L1ZQGe3RJH9q5bM2GWt:-L1ZQGe3RJH9q5bM2GWu:b-8gkfnk) ([raw file](https://github.com/cockroachdb/cockroach/blob/79968b95c10bcd189226e7fcb219dd22d3c08da7/pkg/storage/store.go#L1067)):*
> ```Go
> 		}
> 
> 		log.Infof(ctx, "calling for %d %s %+v", rangeID, unsafeKey.Key, msg)
> ```

Remove this before merging?

---

*[pkg/storage/store.go, line 1149 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/21120#-L1ZQgM--UtFYbAaho48:-L1ZQgM--UtFYbAaho49:bi3zboo) ([raw file](https://github.com/cockroachdb/cockroach/blob/79968b95c10bcd189226e7fcb219dd22d3c08da7/pkg/storage/store.go#L1149)):*
> ```Go
> 	}
> 
> 	// Write a final bogus batch so that we get to do a sync commit, which
> ```

Instead of a bogus batch, write a "migrated" flag so we don't have to repeat this scan on every startup. 

---

*[pkg/storage/store_test.go, line 309 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/21120#-L1ZSP4OdK7zRRpMA2oP:-L1ZSP4PeCFeb88ziaKj:bkmi0x5) ([raw file](https://github.com/cockroachdb/cockroach/blob/79968b95c10bcd189226e7fcb219dd22d3c08da7/pkg/storage/store_test.go#L309)):*
> ```Go
> 	if diff := pretty.Diff(wanted, seen); len(diff) > 0 {
> 		pretty.Ldiff(t, wanted, seen)
> 		t.Fatal("diff(wanted, seen) is nontrivial")
> ```

s/trivial/empty/

---

*[pkg/storage/store_test.go, line 329 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/21120#-L1ZR2aODIrkatHAPllv:-L1ZR2aODIrkatHAPllw:bqqjcqi) ([raw file](https://github.com/cockroachdb/cockroach/blob/79968b95c10bcd189226e7fcb219dd22d3c08da7/pkg/storage/store_test.go#L329)):*
> ```Go
> 		// GC'ed replicas.
> 		keys.RaftTombstoneIncorrectLegacyKey(2 /* rangeID */),
> 		keys.RaftTombstoneIncorrectLegacyKey(3 /* rangeID */),
> ```

The test should include some non-consecutive range IDs and maybe some non-tombstone keys. 

Have the comments rotted? I don't see what's "Existing" or "GC'ed" here. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21120#-:-L1ZSW32_jOs807dwl63:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
I realized there's a migration concern that I overlooked. By immediately rewriting the tombstones when the binary starts, downgrading to the previous version (which then doesn't know about the new tombstones) is incorrect. One option is to let v2.0 always use both tombstones and then to have v2.1 switch over to only the new, but I don't like that this means fixing the stats one-offs for half a year.
Instead, I think I'll have to postpone the rewriting until after the cluster version is bumped to `v2.0`. Mildly annoying, but doable. Nodes always look for both tombstones, and write the old one unless they know about the bumped cluster version (in which case they write the new one only). The startup sequence rewrite isn't actually needed in v2.0 but only in v2.1 (and there, only once), but I don't think postponing putting in the code is a good idea, so my suggestion would be to have it work much like now (rewrite on every start conditional on a good enough persisted cluster version), if that works for @bdarnell. Otherwise I can gate it behind a `binary > 2.0` check and make it fire-once, but I don't like having some surprise thing that fires at some point in the future, when perhaps we've all but forgotten about it.

---

Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/storage/store.go, line 1053 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/21120#-L1ZP5GFcXD1BZQXA9vd:-L1dYFCpOxOSn26THL0V:b353b8a) ([raw file](https://github.com/cockroachdb/cockroach/blob/79968b95c10bcd189226e7fcb219dd22d3c08da7/pkg/storage/store.go#L1053)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This looks like it might skip keys: If we have

- /Local/Range/1/u/Tombstone
- /Local/Range/3/u/Something
- /Local/Range/3/u/Tombstone

then when we try to seek to 2/u/Tombstone, we'll go to 3/u/Something instead. We have to seek to 3/u/Tombstone, but this increment will send us to 4/u/Tombstone instead. We should only change rangeID once, either to increment it or to set it to the one that we got by decoding.

(This may actually be safe since we should only have a tombstone when nothing else exists for the range, but I'd feel better with this change)
</blockquote></details>

Good point. I was disappointed initially that randomized testing didn't exercise that bug. When I looked histories that *should* and didn't, I realized there was actually another bug just above that made the behavior correct: that first rangeID-bumping path was never actually taken (due to using `mvccKey` instead of `unsafeKey`).

---

*[pkg/storage/store.go, line 1057 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/21120#-L1ZQ5TBMtNkEmqRYXCh:-L1_Q3iW2YFNJXCx-SbP:bg01c0k) ([raw file](https://github.com/cockroachdb/cockroach/blob/79968b95c10bcd189226e7fcb219dd22d3c08da7/pkg/storage/store.go#L1057)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

We should explicitly clear `msg` on each cycle; otherwise bits from one iteration could bleed into the next (not an issue for the trivial RaftTruncatedState but could matter for other protos)
</blockquote></details>

`MVCCGetProto` already resets the message.

---

*[pkg/storage/store.go, line 1067 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/21120#-L1ZQGe3RJH9q5bM2GWt:-L1_OrN82rNQva4CBzx4:b-a4wcvi) ([raw file](https://github.com/cockroachdb/cockroach/blob/79968b95c10bcd189226e7fcb219dd22d3c08da7/pkg/storage/store.go#L1067)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Remove this before merging?
</blockquote></details>

Oops. Done.

---

*[pkg/storage/store_test.go, line 309 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/21120#-L1ZSP4OdK7zRRpMA2oP:-L1_PG-vDp_yk6Cghl2R:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/79968b95c10bcd189226e7fcb219dd22d3c08da7/pkg/storage/store_test.go#L309)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

s/trivial/empty/
</blockquote></details>

Done.

---

*[pkg/storage/store_test.go, line 329 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/21120#-L1ZR2aODIrkatHAPllv:-L1_PTdsG8iUOiX9Ss7m:b-6fhi7q) ([raw file](https://github.com/cockroachdb/cockroach/blob/79968b95c10bcd189226e7fcb219dd22d3c08da7/pkg/storage/store_test.go#L329)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

The test should include some non-consecutive range IDs and maybe some non-tombstone keys. 

Have the comments rotted? I don't see what's "Existing" or "GC'ed" here. 
</blockquote></details>

What bug is that supposed to catch? The bulk of the work is done by `IterateIDPrefixKeys` and that's tested extensively.

Comments hadn't rotted, but i clarified them.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21120#-:-L1d_NhN2gdVUzZZ8GHv:b-ip00y1)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
D'oh! There are no simple migrations... :(

Your plan SGTM.

---

Reviewed 3 of 5 files at r5, 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/storage/store.go, line 1057 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/21120#-L1ZQ5TBMtNkEmqRYXCh:-L1hDsf9msCbQszJ79Wp:b6nc6bh) ([raw file](https://github.com/cockroachdb/cockroach/blob/79968b95c10bcd189226e7fcb219dd22d3c08da7/pkg/storage/store.go#L1057)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

`MVCCGetProto` already resets the message.
</blockquote></details>

OK, I was confused by the difference between protoutil.Unmarshal and msg.Unmarshal.

---

*[pkg/storage/store_test.go, line 329 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/21120#-L1ZR2aODIrkatHAPllv:-L1hCmPIpqa8ZyePWxdT:b-v2cv5i) ([raw file](https://github.com/cockroachdb/cockroach/blob/79968b95c10bcd189226e7fcb219dd22d3c08da7/pkg/storage/store_test.go#L329)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

What bug is that supposed to catch? The bulk of the work is done by `IterateIDPrefixKeys` and that's tested extensively.

Comments hadn't rotted, but i clarified them.
</blockquote></details>

It's redundant, but changing the range IDs would be a free way to get a little more end-to-end coverage. Adding other keys probably isn't worth it. (I originally wrote this comment as I was reading through this file backwards)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21120#-:-L1hB_Cd-hvObMC9BPcP:b-c55739)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Updated, PTAL!

---

Review status: 4 of 14 files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21120#-:-L1ssJUPOxJ-vlekbC6y:b-rr168o)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 10 of 12 files at r7, 2 of 2 files at r8.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica.go, line 847 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/21120#-L1y6e5a7sfcLPcPsNM7:-L1y6e5a7sfcLPcPsNM8:b-r0bh0d) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb2789be65d87325f1724f78ddc8de466b8931c0/pkg/storage/replica.go#L847)):*
> ```Go
> 
> 	tombstoneKey := keys.RaftTombstoneKey(desc.RangeID)
> 	if !r.store.cfg.Settings.Version.IsMinSupported(cluster.VersionUnreplicatedTombstoneKey) {
> ```

Couldn't this race with the migration? You could write a new legacy key after the migration has run, or write a new tombstone what will be overwritten by an old one in the migration. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21120)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica.go, line 847 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/21120#-L1y6e5a7sfcLPcPsNM7:-L1yD6wgD_9zofXt8vm6:b-1m7jig) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb2789be65d87325f1724f78ddc8de466b8931c0/pkg/storage/replica.go#L847)):*
> Couldn't this race with the migration

That's correct, and I hope I didn't say anywhere that it couldn't. This race is the reason the legacy tombstone can be retired only in > v2.1, when the migration into 2.1 reliably clears out any race-created legacy tombstones.

You're right though that the migration needs to make sure it isn't "winding back the clock". I don't think RocksDB provides sufficient primitives here, so my solution would be to drop the callback and run the migration at start time when the persisted cluster version is `>= v2.0`. We'll still be freed from legacy tombstones when the v2.1 binary runs, and the concurrency concerns go away.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21120)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica.go, line 847 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/21120#-L1y6e5a7sfcLPcPsNM7:-L1zDcqGTskiglftOh3n:bdfiu5u) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb2789be65d87325f1724f78ddc8de466b8931c0/pkg/storage/replica.go#L847)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

> Couldn't this race with the migration

That's correct, and I hope I didn't say anywhere that it couldn't. This race is the reason the legacy tombstone can be retired only in > v2.1, when the migration into 2.1 reliably clears out any race-created legacy tombstones.

You're right though that the migration needs to make sure it isn't "winding back the clock". I don't think RocksDB provides sufficient primitives here, so my solution would be to drop the callback and run the migration at start time when the persisted cluster version is `>= v2.0`. We'll still be freed from legacy tombstones when the v2.1 binary runs, and the concurrency concerns go away.
</blockquote></details>

Yeah, the problem is just the migration moving things back in time. Moving this to startup SGTM. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21120)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Updated, PTAL! The changes here are a) making the migration properly take into account existing new-style tombstones, b) calling the migration only on startup with a high enough persisted cluster version, and 3) improving test coverage somewhat.

I had to rebase against my recent consistency checker refactors, so a bit of scrutiny is appropriate.

---

Review status: 0 of 16 files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/storage/store_test.go, line 363 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/21120#-L2LdwVprXFCkitPYjT7:-L2LdwVprXFCkitPYjT8:b3m1iz3) ([raw file](https://github.com/cockroachdb/cockroach/blob/5650ec60ec340fe4737c1c1ac797ffdbef0742a5/pkg/storage/store_test.go#L363)):*
> // Put down a few fake legacy tombstones (#12154) to jog `migrateLegacyTombstones`, including
> 		// one for the existing range, but also some for non-existing ones.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21120#-:-L2LuKkhAlecoI8T_ttU:b3dg52g)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 17 files at r9, 15 of 15 files at r10.
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/server/node.go, line 427 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/21120#-L2Mwpy0dOn-pDt9NTSP:-L2Mwpy0dOn-pDt9NTSQ:b5j49wd) ([raw file](https://github.com/cockroachdb/cockroach/blob/14b0a0835e74584648c6b8cc7d889f367934aba4/pkg/server/node.go#L427)):*
> ```Go
> 	n.startComputePeriodicMetrics(n.stopper, DefaultMetricsSampleInterval)
> 	// Be careful about moving this line above `startStores`; store migrations rely
> 	// on the fact that the cluster version has not been updated via Gossip.
> ```

I don't understand what this comment means. Why would the migration fail if this was moved earlier?

---

*[pkg/storage/store.go, line 1184 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/21120#-L2MvlJgd-gi1S4LAgVD:-L2MvlJhmMBY9EKVfmQv:b-qtammv) ([raw file](https://github.com/cockroachdb/cockroach/blob/14b0a0835e74584648c6b8cc7d889f367934aba4/pkg/storage/store.go#L1184)):*
> ```Go
> 	// Migrate legacy tombstones away if we can. We want to do this when the node boots with a
> 	// cluster version as below or higher, i.e. not when the upgrade happens for a running node.
> 	// In fact, we do it on *every* such boot (i.e. not only once); this is because we want the
> ```

Comment that we don't actually need to do this in 2.0 and could wait to do the migration until the first boot with a 2.1 binary. We're just running it here to ensure that the code doesn't rot between now and then (and we could turn it off in 2.0 releases if it's a performance issue). 

---

*[pkg/testutils/error.go, line 66 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/21120#-L2MwBTKWMFd7yHJK2-5:-L2MwBTKWMFd7yHJK2-6:bq18fdf) ([raw file](https://github.com/cockroachdb/cockroach/blob/14b0a0835e74584648c6b8cc7d889f367934aba4/pkg/testutils/error.go#L66)):*
> ```Go
> 	// Don't forget to update the corresponding test when making adjustments
> 	// here.
> 	return IsError(err, "(connection reset by peer|connection refused|failed to send RPC|EOF|result is ambiguous|connection is unavailable)")
> ```

Why is this changing now?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21120#-:-L2Mx5rS8TUo5wqwWE5G:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tbg committed:
storage: make RaftTombstoneKey unreplicated

RaftTombstoneKey was accidentally made a replicated key when it was first
introduced, a problem we first realized existed when it was [included in
snapshots].

At the time, we included workarounds to skip this key in various places
(snapshot application, consistency checker) but of course we have failed to
insert further hacks of the same kind elsewhere since (the one that prompting
this PR being the stats recomputation on splits, which I'm looking into as part
of #20181 -- unfortunately this commit doesn't seem to pertain to that problem)

It feels sloppy that we didn't follow through back then, but luckily the damage
appears to be limited; it is likely that the replicated existence of this key
results in MVCCStats SysBytes inconsistencies, but as it happens, these stats
are [already] [very] [inconsistent].

This commit does a few things:

- renames the old tombstone key to `RaftIncorrectLegacyTombstoneKey`
- introduces a (correctly unreplicated) `RaftTombstoneKey`
- introduces a migration. Once activated, only the new tombstone is written,
  but both tombstones are checked. Additionally, as the node restarts, all
  legacy tombstones are replaced by correct non-legacy ones.
- when applying a snapshot, forcibly delete any legacy tombstone contained
  within (even before the cluster version is bumped). This prevents new
  legacy tombstones to trickle on from other nodes.

`RaftIncorrectLegacyTombstoneKey` can be purged from the codebase in binaries
post v2.1, as at that point all peers have booted with a version that runs the
migration.

Thus, post v2.1, the replica consistency checker can stop skipping the legacy
tombstone key.

Fixes #12154.

Release note: None

[included in snapshots]: https://github.com/cockroachdb/cockroach/pull/12131
[already]: https://github.com/cockroachdb/cockroach/issues/20554
[very]: https://github.com/cockroachdb/cockroach/pull/20996
[inconsistent]: https://github.com/cockroachdb/cockroach/pull/21070

tschottdorf commented:
TFTR!

---

Review status: 11 of 15 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[pkg/server/node.go, line 427 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/21120#-L2Mwpy0dOn-pDt9NTSP:-L2QndCtE4Aq4DyE9ajv:bvilvxr) ([raw file](https://github.com/cockroachdb/cockroach/blob/14b0a0835e74584648c6b8cc7d889f367934aba4/pkg/server/node.go#L427)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I don't understand what this comment means. Why would the migration fail if this was moved earlier?
</blockquote></details>

It wouldn't fail, but it would run when we wouldn't want it to (in a probably rare race). I clarified the comment.

---

*[pkg/storage/store.go, line 1184 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/21120#-L2MvlJgd-gi1S4LAgVD:-L2Qo4JQd9O_iNRzH7WG:bpyq23q) ([raw file](https://github.com/cockroachdb/cockroach/blob/14b0a0835e74584648c6b8cc7d889f367934aba4/pkg/storage/store.go#L1184)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Comment that we don't actually need to do this in 2.0 and could wait to do the migration until the first boot with a 2.1 binary. We're just running it here to ensure that the code doesn't rot between now and then (and we could turn it off in 2.0 releases if it's a performance issue). 
</blockquote></details>

Done. I also log a message at `Info` if the migration takes longer than 10s so that we'll know when to act.

---

*[pkg/testutils/error.go, line 66 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/21120#-L2MwBTKWMFd7yHJK2-5:-L2QokdP7br2CptCIr5W:brgy6bl) ([raw file](https://github.com/cockroachdb/cockroach/blob/14b0a0835e74584648c6b8cc7d889f367934aba4/pkg/testutils/error.go#L66)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Why is this changing now?
</blockquote></details>

I think I just happened to catch the flake. Removing this commit so you can look at it in its own PR. The actual message probably changed with a gRPC bump.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21120#-:-L2QpRC8sMUHez5XAVmK:b-814g8k)*
<!-- Sent from Reviewable.io -->
